latch_cli/snakemake/config/utils.py

```diff
@@ -0,0 +1,109 @@
+from dataclasses import asdict, fields, is_dataclass, make_dataclass
+from enum import Enum
+from typing import Any, Dict, List, Optional, Type, Union, get_args, get_origin
+
+from typing_extensions import TypeAlias, TypeGuard, TypeVar
+
+from latch.types.directory import LatchDir
+from latch.types.file import LatchFile
+from latch_cli.utils import identifier_suffix_from_str
+
+JSONValue: TypeAlias = Union[int, str, bool, float, List["JSONValue"], "JSONDict"]
+JSONDict: TypeAlias = Dict[str, "JSONValue"]
+
+
+def parse_type(v: JSONValue, name: Optional[str] = None) -> Type:
+    if is_primitive_value(v):
+        return type(v)
+
+    if isinstance(v, list):
+        parsed_types = tuple(parse_type(x) for x in v)
+        return List[Union[parsed_types]]
+
+    assert isinstance(v, dict)
+
+    if name is None:
+        name = "SnakemakeRecord"
+
+    fields = {}
+    for k, x in v.items():
+        fields[identifier_suffix_from_str(k)] = parse_type(
+            x, identifier_suffix_from_str(k)
+        )
+
+    return make_dataclass(name, fields.items())
+
+
+def is_primitive_type(
+    typ: Type,
+) -> TypeGuard[Union[Type[str], Type[bool], Type[int], Type[float]]]:
+    return _is_primitive(t=typ)
+
+
+def is_primitive_value(
+    val: Any,
+) -> TypeGuard[Union[str, bool, int, float]]:
+    return _is_primitive(v=val)
+
+
+def _is_primitive(
```

Comment:
kinda weird? why not
```py3
def is_primitive_value(x: object):
	return is_primitive_type(type(x))
```
instead?

btw `object` type does not work for some things but is the closest thing to `unknown`